staging/src/k8s.io/apiserver/pkg/endpoints/installer.go

```diff
@@ -867,8 +855,19 @@ func (a *APIInstaller) registerResourceHandlers(path string, storage rest.Storag
 		apiResource.Verbs = append(apiResource.Verbs, kubeVerb)
 	}
 	sort.Strings(apiResource.Verbs)
-	apiResource.ShortNames = shortNames
-	apiResource.Categories = categories
+
+	if shortNamesProvider, ok := storage.(rest.ShortNamesProvider); ok {
+		apiResource.ShortNames = shortNamesProvider.ShortNames()
+	}
+	if categoriesProvider, ok := storage.(rest.CategoriesProvider); ok {
+		apiResource.Categories = categoriesProvider.Categories()
+	}
+	if gvkProvider, ok := storage.(rest.GroupVersionKindProvider); ok {
+		gvk := gvkProvider.GroupVersionKind()
+		apiResource.Group = gvk.Group
+		apiResource.Version = gvk.Version
+		apiResource.Kind = gvk.Kind
```

Comment:
Now we have a KindProvider and a GroupVersionKindProvider provider. Wouldn't it make things clearer to have a GroupVersionProvider to complement the KindProvider?

Reply:
> Now we have a KindProvider and a GroupVersionKindProvider provider. Wouldn't it make things clearer to have a GroupVersionProvider to complement the KindProvider?

The other is for federation, I'd rather not change it for 1.8 since the intent is different.

Reply:
What do you mean with federation? Compare https://github.com/deads2k/kubernetes/blob/7b44627d0372a3febbe906441c28f43323760bf9/staging/src/k8s.io/apiserver/pkg/endpoints/installer.go#L374. Isn't that the source of a custom kind? Your interface does gv in addition.

Reply:
> What do you mean with federation? Compare https://github.com/deads2k/kubernetes/blob/7b44627d0372a3febbe906441c28f43323760bf9/staging/src/k8s.io/apiserver/pkg/endpoints/installer.go#L374. Isn't that the source of a custom kind? Your interface does gv in addition.

It's only implemented by federation types to lie about their kinds during discovery.  Every interaction with federation reveals dragons that I'd prefer to slay early in 1.9 as opposed to late in 1.8.

Reply:
Ic, then just add a comment to your new interface that it trumps over the KindProvider. After that lgtm.

Reply:
> Ic, then just add a comment to your new interface that it trumps over the KindProvider. After that lgtm.

done